transport.go

```diff
@@ -0,0 +1,409 @@
+		return nil, err
+	}
+	conf = populateConfig(conf)
+	if err := t.init(false); err != nil {
+		return nil, err
+	}
+	var onClose func()
+	if t.isSingleUse {
+		onClose = func() { t.Close() }
+	}
+	return dial(ctx, newSendConn(t.conn, addr, nil), t.ConnectionIDGenerator, t.handlerMap, tlsConf, conf, onClose, false)
+}
+
+// DialEarly dials a new connection, attempting to use 0-RTT if possible.
+func (t *Transport) DialEarly(ctx context.Context, addr net.Addr, tlsConf *tls.Config, conf *Config) (EarlyConnection, error) {
+	if err := validateConfig(conf); err != nil {
+		return nil, err
+	}
+	conf = populateConfig(conf)
+	if err := t.init(false); err != nil {
+		return nil, err
+	}
+	var onClose func()
+	if t.isSingleUse {
+		onClose = func() { t.Close() }
+	}
+	return dial(ctx, newSendConn(t.conn, addr, nil), t.ConnectionIDGenerator, t.handlerMap, tlsConf, conf, onClose, true)
+}
+
+func setReceiveBuffer(c net.PacketConn, logger utils.Logger) error {
+	conn, ok := c.(interface{ SetReadBuffer(int) error })
+	if !ok {
+		return errors.New("connection doesn't allow setting of receive buffer size. Not a *net.UDPConn?")
+	}
+	size, err := inspectReadBuffer(c)
+	if err != nil {
+		return fmt.Errorf("failed to determine receive buffer size: %w", err)
+	}
+	if size >= protocol.DesiredReceiveBufferSize {
+		logger.Debugf("Conn has receive buffer of %d kiB (wanted: at least %d kiB)", size/1024, protocol.DesiredReceiveBufferSize/1024)
+		return nil
+	}
+	if err := conn.SetReadBuffer(protocol.DesiredReceiveBufferSize); err != nil {
+		return fmt.Errorf("failed to increase receive buffer size: %w", err)
+	}
+	newSize, err := inspectReadBuffer(c)
+	if err != nil {
+		return fmt.Errorf("failed to determine receive buffer size: %w", err)
+	}
+	if newSize == size {
+		return fmt.Errorf("failed to increase receive buffer size (wanted: %d kiB, got %d kiB)", protocol.DesiredReceiveBufferSize/1024, newSize/1024)
+	}
+	if newSize < protocol.DesiredReceiveBufferSize {
+		return fmt.Errorf("failed to sufficiently increase receive buffer size (was: %d kiB, wanted: %d kiB, got: %d kiB)", size/1024, protocol.DesiredReceiveBufferSize/1024, newSize/1024)
+	}
+	logger.Debugf("Increased receive buffer size to %d kiB", newSize/1024)
+	return nil
+}
+
+// only print warnings about the UDP receive buffer size once
+var receiveBufferWarningOnce sync.Once
+
+func (t *Transport) init(isServer bool) error {
+	t.initOnce.Do(func() {
+		getMultiplexer().AddConn(t.Conn)
+
+		conn, err := wrapConn(t.Conn)
+		if err != nil {
+			t.initErr = err
+			return
+		}
+
+		t.logger = utils.DefaultLogger // TODO: make this configurable
+		t.conn = conn
+		t.handlerMap = newPacketHandlerMap(t.StatelessResetKey, t.enqueueClosePacket, t.logger)
+		t.listening = make(chan struct{})
+
+		t.closeQueue = make(chan closePacket, 4)
+
+		if t.ConnectionIDGenerator != nil {
+			t.connIDLen = t.ConnectionIDGenerator.ConnectionIDLen()
+		} else {
+			connIDLen := t.ConnectionIDLength
```

Comment:
If `(Transport).ConnectionIDLength` was removed, this branch on `t.isSingleUse` and `isServer` can be eliminated.